### PR TITLE
Hooks handler

### PR DESCRIPTION
**Pull request by mariajgrimaldi**
This PR adds a signal receiver function that checks the configuration to call a defined task. Then, saves the results in an audit model

The configuration looks like the following:

```
{
"EOX_HOOKS_PRE_ENROLLMENT_SIGNAL":
   {
     "action": "eox_hooks.actions.default_actions",
     "module": "default_actions",
     "timing": True,
     "sync": True,
   }
}
```

### CODE REVIEW

File: eox_hooks/receivers.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+from eox_hooks.constants import Status
+from eox_hooks.models import HookExecutionAudit
+
+log = logging.getLogger(__name__)
+
+DEFAULT_TASK_MODULE = "eox_hooks.tasks"
+DEFAULT_TASK_NAME = "default_task"
+
+
+def hooks_handler(sender, **kwargs):
```
Comment (andrey-canon) [inline eox_hooks/receivers.py:16]: is this mean that we will have only **one** receiver method ???

File: eox_hooks/receivers.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    """Eox-hooks signal receiver."""
+    if not getattr(settings, "USE_EOX_HOOKS", False):
+        return
+
+    try:
+        current_sender = sender.__name__
```
Comment (andrey-canon) [inline eox_hooks/receivers.py:22]: do you have any example when the sender does not have` __name__` ?

File: eox_hooks/receivers.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+def __task_executer(sender, configuration, **kwargs):
+    """
+    Function that executes tasks given a specific configuration.
+
+    This function also starts the timer for the timing feature and set the tasks as async/sync.
+    This given the current configuration.
+    """
+    task = __task_lookup(configuration.get("task"))
+    try:
+        task(sender=sender, **kwargs)
```
Comment (andrey-canon) [inline eox_hooks/receivers.py:43]: I assume this should be a celery task and this must be task.delay or task.apply_async

File: eox_hooks/receivers.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        audit_entry = HookExecutionAudit.objects.create(
+            status=Status.FAIL,
+            hook_name=sender,
+            task_name=task.__name__,
+            tenant_domain="tenant_domain"
+        )
+        log.warning(str(audit_entry))
+
+
+def __task_lookup(path):
```
Comment (andrey-canon) [inline eox_hooks/receivers.py:61]: why double underscore ?

File: eox_hooks/receivers.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    """
+    try:
+        module_name, current_task = path.rsplit(".", 1)
+    except AttributeError:
+        module_name, current_task = DEFAULT_TASK_MODULE, DEFAULT_TASK_NAME
+
+    try:
+        module = import_module(module_name)
+        return getattr(module, current_task)
+    except (ImportError, AttributeError):
```
Comment (andrey-canon) [inline eox_hooks/receivers.py:77]: you can handle every exception different in order to write specific messages, importerror - wrong module, attribute-wrong task name

File: eox_hooks/receivers.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+def __task_lookup(path):
+    """
+    Function that returns a function defined in path, if the path is not find then a default task is
+    returned.
+
+    The path must look like: eox_hooks.tasks.default_logging
+    Where eox_hooks.tasks is module_name and task_name is default_logging
+    """
+    try:
+        module_name, current_task = path.rsplit(".", 1)
```
Comment (andrey-canon) [inline eox_hooks/receivers.py:70]: imo it is better if  the method has the parameters module and task name

File: eox_hooks/receivers.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def __task_executer(sender, configuration, **kwargs):
+    """
+    Function that executes tasks given a specific configuration.
+
+    This function also starts the timer for the timing feature and set the tasks as async/sync.
+    This given the current configuration.
+    """
+    task = __task_lookup(configuration.get("task"))
```
Comment (andrey-canon) [inline eox_hooks/receivers.py:41]: the configuration could be a dict instead of a string

File: eox_hooks/receivers.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        )
+        log.info(str(audit_entry))
+    except Exception:  # pylint: disable=broad-except
+        audit_entry = HookExecutionAudit.objects.create(
+            status=Status.FAIL,
+            hook_name=sender,
+            task_name=task.__name__,
+            tenant_domain="tenant_domain"
+        )
+        log.warning(str(audit_entry))
```
Comment (andrey-canon) [inline eox_hooks/receivers.py:58]: log.error?

File: eox_hooks/receivers.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    try:
+        task(sender=sender, **kwargs)
+        audit_entry = HookExecutionAudit.objects.create(
+            status=Status.SUCCESS,
+            hook_name=sender,
+            task_name=task.__name__,
+            tenant_domain="tenant_domain"
+        )
+        log.info(str(audit_entry))
+    except Exception:  # pylint: disable=broad-except
```
Comment (andrey-canon) [inline eox_hooks/receivers.py:51]: the async tasks don't raise exceptions

File: eox_hooks/apps.py [hunk 10]
```diff
@@ -38,5 +38,16 @@ class EoxHooksConfig(AppConfig):
...
             },
-        }
+        },
+        'signals_config': {
+            'lms.djangoapp': {
+                'relative_path': 'receivers',
+                'receivers': [
+                    {
+                        'receiver_func_name': 'hooks_handler',
+                        'signal_path': 'django.core.signals.request_started',
```
Comment (andrey-canon) [inline eox_hooks/apps.py:48]: I think this is a bad idea, because this will be executed in every requests and when we need to change that we hace to create a new plugin version, have you tried with settings ??'

File: eox_hooks/receivers.py (hunk 1 above)
Comment (mariajgrimaldi) [reply eox_hooks/receivers.py:16]: yes! that's the plan

File: eox_hooks/apps.py (hunk 10 above)
Comment (mariajgrimaldi) [reply eox_hooks/apps.py:48]: I was testing with the platform and left that there, sorry. I'll erase the signal_path

File: eox_hooks/receivers.py (hunk 5 above)
Comment (mariajgrimaldi) [reply eox_hooks/receivers.py]: Ok that would be better

File: eox_hooks/receivers.py (hunk 6 above)
Comment (mariajgrimaldi) [reply eox_hooks/receivers.py]: Yeah, that could be better

File: eox_hooks/receivers.py (hunk 2 above)
Comment (mariajgrimaldi) [reply eox_hooks/receivers.py:20]: I'll add it

File: eox_hooks/tasks_handler.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+
+def task_lookup(module_name, task):
+    """
+    Function that returns a function defined in path, if the path is not find then a default task is
+    returned.
+    The path must look like: eox_hooks.tasks.default_logging
+    Where eox_hooks.tasks is module_name and task_name is default_logging
+    """
+    try:
```
Comment (andrey-canon) [inline eox_hooks/tasks_handler.py:40]: image

File: eox_hooks/tests/test_receivers.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    SingalsTestCase: Test Receivers.
+"""
+from django.test import TestCase, override_settings
+from mock import Mock, patch
+
+from eox_hooks.receivers import hooks_handler
+
+
+@override_settings(
+    EOX_HOOKS_DEFINITIONS={
```
Comment (andrey-canon) [inline eox_hooks/tests/test_receivers.py:13]: Why is this here?, if there is a file for this kind a things

File: eox_hooks/tasks_handler.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        module_name, task = configuration.get("task").rsplit(".", 1)
+    except AttributeError:
+        module_name, task = DEFAULT_TASK_MODULE, DEFAULT_TASK_NAME
+
+    task = task_lookup(module_name, task)
+    try:
+        task(sender=sender, **kwargs)
+        audit_task_execution(Status.SUCCESS, sender, task)
+    except Exception:  # pylint: disable=broad-except
```
Comment (andrey-canon) [inline eox_hooks/tasks_handler.py:29]: tasks don't raise exceptions

File: eox_hooks/tasks_handler.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    """
+    try:
+        module_name, task = configuration.get("task").rsplit(".", 1)
+    except AttributeError:
+        module_name, task = DEFAULT_TASK_MODULE, DEFAULT_TASK_NAME
+
+    task = task_lookup(module_name, task)
+    try:
+        task(sender=sender, **kwargs)
+        audit_task_execution(Status.SUCCESS, sender, task)
```
Comment (andrey-canon) [inline eox_hooks/tasks_handler.py:28]: this doesn't make any sense with async tasks

File: eox_hooks/models.py [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    status = models.PositiveIntegerField(
+        choices=Status.choices(), default=Status.SUCCESS, editable=False,
+    )
+    executed_at = models.DateTimeField(auto_now_add=True, editable=False)
+    hook_name = models.CharField(max_length=150)
+    task_name = models.CharField(max_length=150)
+    tenant_domain = models.CharField(max_length=150)
+    task_duration = models.FloatField(null=True)
+
+    objects = models.Manager()
```
Comment (andrey-canon) [inline eox_hooks/models.py:30]: please try with this https://stackoverflow.com/a/1416297

File: eox_hooks/tests/test_tasks_handler.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+Classes:
+    SingalsTestCase: TestTasksHandler.
+"""
+from django.test import TestCase, override_settings
+from mock import patch
+
+from eox_hooks.tasks_handler import task_handler
+
+
+@override_settings(USE_EOX_HOOKS=True)
```
Comment (andrey-canon) [inline eox_hooks/tests/test_tasks_handler.py:12]: x2

File: eox_hooks/actions_handler.py [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+        return getattr(module, action)
+    except ImportError:
+        log.info("The module {} does not exist with the action {}. A default action will be used."
+                 .format(module_name, action))
+        module, action = import_module(DEFAULT_ACTIONS_MODULE), DEFAULT_ACTIONS_NAME
+    except AttributeError:
+        log.info("The action {} does not exist in the module {}. A default action will be used."
+                 .format(action, module.__name__))
+        module, action = import_module(DEFAULT_ACTIONS_MODULE), DEFAULT_ACTIONS_NAME
+        return getattr(module, action)
```
Comment (andrey-canon) [inline eox_hooks/actions_handler.py:46]:
```suggestion
    except ImportError:
        message = "The module {} does not exist with the action {}. A default action will be used.".format(
            module_name,
            action,
        )
    except AttributeError:
        message = "The action {} does not exist in the module {}. A default action will be used.".format(
            action,
            module.__name__,
        )

    log.warning(message)

    return custom_action_mock
```

File: eox_hooks/actions_handler.py [hunk 18]
```diff
@@ -0,0 +1,46 @@
+"""This module defines the handlers for the actions triggered by a signal received by eox-hooks."""
+import logging
+from importlib import import_module
+
```
Comment (andrey-canon) [inline eox_hooks/actions_handler.py:4]:
```suggestion
from eox_hooks.tests.test_utils import  custom_action_mock
```
This will work with future suggestions, also you can delete DEFAULT_ACTIONS_MODULE and DEFAULT_ACTION_NAME

File: eox_hooks/actions_handler.py [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+def action_handler(sender, configuration, **kwargs):
+    """
+    Function that executes actions given a specific configuration.
+    """
+    try:
+        module_name, action = configuration.get("action").rsplit(".", 1)
+    except AttributeError:
+        module_name, action = DEFAULT_ACTIONS_MODULE, DEFAULT_ACTIONS_NAME
+
+    action = action_lookup(module_name, action)
```
Comment (andrey-canon) [inline eox_hooks/actions_handler.py:20]:
```suggestion
    action = action_lookup(
        action=configuration.get("action", ""),
        module_name=configuration.get("module", "missing_module"),
    )

```

File: eox_hooks/tests/test_receivers.py (hunk 12 above)
Comment (mariajgrimaldi) [reply eox_hooks/tests/test_receivers.py:13]: changed

File: eox_hooks/actions_handler.py [hunk 20]
```diff
@@ -0,0 +1,46 @@
+"""This module defines the handlers for the actions triggered by a signal received by eox-hooks."""
+import logging
+from importlib import import_module
+
+log = logging.getLogger(__name__)
+
+DEFAULT_ACTIONS_MODULE = "eox_hooks.tests.test_utils"
```
[the remark this answers is not in the document]
Comment (mariajgrimaldi) [reply eox_hooks/actions_handler.py:7]: done

File: eox_hooks/actions_handler.py [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+DEFAULT_ACTIONS_NAME = "custom_action_mock"
```
[the remark this answers is not in the document]
Comment (mariajgrimaldi) [reply eox_hooks/actions_handler.py:8]: changed

File: eox_hooks/tests/test_receivers.py [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+class TestReceivers(TestCase):
+    """Receivers test class."""
+
+    def setUp(self):
+        """Set up class for Receivers testing."""
+        self.sender = Mock()
+        self.trigger_event = "sender.example"
+
+    @patch("eox_hooks.receivers.action_handler")
+    def test_trigger_response_action(self, custom_action):
```
Comment (andrey-canon) [inline eox_hooks/tests/test_receivers.py:22]:
```suggestion
    def test_trigger_response_action(self, action_handler):
```

File: eox_hooks/tests/test_receivers.py [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+        """
+        Used to test that if USE_EOX_HOOKS is set to true then the handler triggers
+        the execution of the action handlers.
+        """
+        custom_action.__name__ = "custom_action_mock"
+
+        hooks_handler(self.sender, trigger_event=self.trigger_event)
+
+        custom_action.assert_called()
```
Comment (andrey-canon) [inline eox_hooks/tests/test_receivers.py:31]:
```suggestion
        custom_action.assert_called_once_with(
                 self.trigger_event,
                 settings.EOX_HOOKS_DEFINITIONS.get(trigger_event, {}),
        )
```

File: eox_hooks/tests/test_receivers.py [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+        """
+        hooks_handler(self.sender, trigger_event=self.trigger_event)
+
+        custom_action.assert_not_called()
+
+    def test_sender_not_in_config(self):
+        """Used to test what happends if the sender is not defined."""
+        self.sender = "non_existant_sender.example"
+
+        with LogCapture():
```
Comment (andrey-canon) [inline eox_hooks/tests/test_receivers.py:48]: check the log string

File: eox_hooks/receivers.py [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+
+
+def hooks_handler(sender, **kwargs):
+    """Eox-hooks signal receiver."""
+    if not getattr(settings, "USE_EOX_HOOKS", False):
+        return
+
+    trigger_event = kwargs.get("trigger_event")
+
+    if trigger_event not in settings.EOX_HOOKS_DEFINITIONS:
```
Comment (felipemontoya) [inline eox_hooks/receivers.py:18]: I suggest we just return silently here. There might be thousands of such calls that dont have any actions configured

Comment (mariajgrimaldi) [reply eox_hooks/receivers.py]: Done! I added:

```
if not settings.EOX_HOOKS_DEFINITIONS or trigger_event not in settings.EOX_HOOKS_DEFINITIONS:
        return
```

File: eox_hooks/tests/test_receivers.py [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+
+    @patch("eox_hooks.receivers.action_handler")
+    def test_trigger_response_action(self, action_handler):
+        """
+        Used to test that if USE_EOX_HOOKS is set to true then the handler triggers
+        the execution of the action handlers.
+        """
+        configuration = settings.EOX_HOOKS_DEFINITIONS.get(self.trigger_event)
+
+        hooks_handler(self.sender, trigger_event=self.trigger_event)
```
Comment (andrey-canon) [inline eox_hooks/tests/test_receivers.py:29]: you have changed the method, this should be with signal instead of trigger event

File: eox_hooks/tests/test_receivers.py [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+    SingalsTestCase: Test Receivers.
+"""
+from django.conf import settings
+from django.test import TestCase, override_settings
+from mock import Mock, patch
+
+from eox_hooks.receivers import hooks_handler
+
+
+class TestReceivers(TestCase):
```
Comment (andrey-canon) [inline eox_hooks/tests/test_receivers.py:13]: it is missing the test case when the dispatch doesn't have a value which starts with `eox-hooks:`

File: eox_hooks/tests/test_actions_handler.py [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+Classes:
+    SingalsTestCase: TestActionHandler.
+"""
+from django.test import TestCase
+from mock import patch
+
+from eox_hooks.actions_handler import action_handler
+
+
+class TestActionHandler(TestCase):
```
Comment (andrey-canon) [inline eox_hooks/tests/test_actions_handler.py:12]: You always test the default case, you could do the following:

1) Create two classes, one for action_handler  and another one for action_lookup
2) (action_handler class) 

       -  Mock action_lookup
       -  test when the value returned by  action_lookup doesn't raise an exception
       -  test when the value returned by  action_lookup raise an exception and silently is True
       -  test when the value returned by  action_lookup raise an exception and silently is False

3) (action_lookup class) 

       -  Mock import_module
       -  test happy path( the module exist and has the attribute)verify the returned value
       -  test when the module exist and doesn't have the attribute, verify the log, and the returned value this should be custom_action_mock
       -  test when import module raises Import error, verify the log and the returned value, this should be custom_action_mock

Comment (mariajgrimaldi) [reply eox_hooks/tests/test_actions_handler.py:12]: Nice suggestion, thank you!

I tried to follow them, the main change I did was using silently because the function that must fail is the one running!

File: eox_hooks/tests/test_actions_handler.py [hunk 29]
```diff
@@ -0,0 +1,112 @@
...
+
+    @patch("eox_hooks.actions_handler.action_lookup")
+    @patch("eox_hooks.tests.test_utils.custom_action_mock")
+    def test_executing_custom_action(self, custom_action_mock, action_lookup_mock):
+        """Used to test executing custom action when hook setting is set to true."""
+        action_lookup_mock.return_value = custom_action_mock
+
+        action_handler(self.trigger_event, self.configuration)
+
+        custom_action_mock.assert_called()
```
Comment (andrey-canon) [inline eox_hooks/tests/test_actions_handler.py:33]:
```suggestion
    @patch("eox_hooks.actions_handler.action_lookup")
    def test_executing_custom_action(self,  action_lookup_mock):
        """Used to test executing custom action when hook setting is set to true."""
        action_mock = Mock()
        action_lookup_mock.return_value = action_mock

        action_handler(self.trigger_event, self.configuration)

        action_lookup_mock.assert_called_once_with(
            module_name=self.configuration.get("module", "missing_module"),
            action=configuration.get("action", ""),
        )
        action_mock.assert_called_once_with(kwargs)
```

Comment (mariajgrimaldi) [reply eox_hooks/tests/test_actions_handler.py:33]: Thank you!!
